### ISSUE TITLE
在brats2020数据集上复现结果，指标有点不一样，想请教一下

### ISSUE THREAD

**Issue by xiao-xinan** [CuCl-2/private_commit#5]
因为没有办法获得RFNet提供的google云盘数据集，我尝试使用仓库中预处理脚本处理原始brats2020数据集，在推理过程继承test.txt文件，具体的推理过程文件如下：

[hetero_training.txt](https://github.com/user-attachments/files/24577384/hetero_training.txt)
上述推理结果不同是由什么导致的？是否是因为数据集预处理不一致导致的？还是因为训练周期等其他问题？

**Comment by CuCl-2** [comment]
- 我知道了，跟数据应该关系不大，epoch有影响但我觉得不是主要因素
- 主要是bath_size，你这里设为了1，这样对比学习就不起作用了
- 如果因为显存不够才设为1的话，我建议权衡crop_size到80，batch_size改成2

**Comment by xiao-xinan** [comment]
感谢您的建议，这对我帮助很大

**Comment by seunghub** [comment]
@CuCl-2  您好，我这边也从 Hugging Face 下载并使用了预训练权重，得到的 log 和 @xiao-xinan 提到的情况比较相似。据我了解，test_loader 的 batch size 应该是固定为 1 的。
不过，与 Hugging Face 页面上提供的 log 相比，目前复现出来的性能有比较明显的差距。因此想请教一下：
1）是否有可能是当前提供的 pretrained checkpoint 存在问题？
2）或者在加载和使用该 checkpoint 时，是否还需要额外调整某些设置或参数？

@xiao-xinan 您好，如果您最终成功复现了论文中的性能，方便分享一下主要做了哪些修改或调整吗？
非常感谢两位的工作与帮助！

**Comment by xiao-xinan** [comment]
> [@CuCl-2](https://github.com/CuCl-2) 您好，我这边也从 Hugging Face 下载并使用了预训练权重，得到的 log 和 [@xiao-xinan](https://github.com/xiao-xinan) 提到的情况比较相似。据我了解，test_loader 的 batch size 应该是固定为 1 的。 不过，与 Hugging Face 页面上提供的 log 相比，目前复现出来的性能有比较明显的差距。因此想请教一下： 1）是否有可能是当前提供的 pretrained checkpoint 存在问题？ 2）或者在加载和使用该 checkpoint 时，是否还需要额外调整某些设置或参数？
> 
> [@xiao-xinan](https://github.com/xiao-xinan) 您好，如果您最终成功复现了论文中的性能，方便分享一下主要做了哪些修改或调整吗？ 非常感谢两位的工作与帮助！

我并没有做其他修改，尝试调整了batch_size后，在测试结果中，指标并没有得到回升。我个人观点是数据集的处理不一样，所以最好的方法应该是重新训练，可惜的是我并没有重新训练了，如果您有尝试重新训练的想法，我也非常荣幸和您一起尝试

**Comment by seunghub** [comment]
> > [@CuCl-2](https://github.com/CuCl-2) 안녕하세요, 저도 Hugging Face에서 제공하는 사전 학습된 가중치를 다운로드하여 사용해 보았는데, [@xiao-xinan](https://github.com/xiao-xinan) 님께서 언급하신 것과 유사한 로그를 얻었습니다. 제가 이해하기로는 `test_loader`의 배치 크기는 1로 고정되어 있어야 합니다. 하지만 Hugging Face 웹사이트에 제공된 로그와 비교했을 때, 제가 재현하고 있는 성능은 상당히 다릅니다. 따라서 다음과 같은 질문을 드리고 싶습니다. 1) 현재 제공되는 사전 학습 체크포인트에 문제가 있는 것일까요? 2) 아니면, 이 체크포인트를 로드하고 사용할 때 추가로 조정해야 할 설정이나 매개변수가 있을까요?
> > [@xiao-xinan](https://github.com/xiao-xinan) 안녕하세요, 논문에 기술된 성능을 성공적으로 재현하셨다면, 어떤 주요 수정이나 조정을 하셨는지 공유해 주시면 감사하겠습니다. 두 분 모두 수고와 도움에 진심으로 감사드립니다!
> 
> 다른 수정은 하지 않았습니다. 배치 크기를 조정해 봤지만 테스트 결과에서 성능 향상은 없었습니다. 제 생각에는 데이터셋 처리 방식이 다르기 때문에 재학습이 최선의 방법일 것 같습니다. 아쉽게도 저는 재학습을 하지 못했습니다. 만약 재학습을 시도해 보고 싶으시다면, 함께 진행해도 좋습니다.


@xiao-xinan 我这边用给定的代码和设置进行过训练，得到的结果似乎和 Hugging Face 上的性能比较接近。因此后来也尝试直接使用 Hugging Face 提供的 pretrained checkpoint，但性能同样没有明显提升。

如果 @CuCl-2 方便的话，是否可以指点一下，为了达到论文中 report 的性能，是否需要对某些部分做额外修改或特别设置？非常感谢！

**Comment by xiao-xinan** [comment]
@CuCl-2 我想向您请教一下，咱们这个模型在brats2018数据集上的指标具体是多少？我尝试继承了您提供的brats2020预训练权重接着训练brats2018数据集200epoch，我发现最后的指标有点高，我感觉可能出现了数据泄露的问题。期待您的回复。

**Comment by CuCl-2** [comment]
To clarify, when I mentioned setting the batch_size to 2, I was referring to the training batch size, not the inference batch size. This is required for contrastive learning to work properly (it does not function when batch_size = 1).

The experiments in the paper were conducted on BraTS2020, and the provided weights can be directly used for inference on BraTS2020.

I did not conduct experiments on BraTS2018. If you want to run experiments on BraTS2018, I recommend training from scratch to avoid potential data leakage.

You can reproduce the results reported in the paper by directly running job_brats.sh without modifying any parameters.

**Comment by CuCl-2** [comment]
The first line of hetero_training.txt in the logs provided on Hugging Face contains the detailed configuration. It is recommended to first reproduce this experiment on BraTS2020, and then transfer it to other datasets or try other configurations.
2025-01-02 14:53:37,571 Namespace(batch_size=2, datapath='/data1/lihaitao/datasets/RFNet/BRATS2020_Training_none_npy', dataname='BRATS2020', savepath='/mnt/data2/lihaitao/output/hetero/Mymodel_RFM/Brats2020_contrastive_112', resume=None, pretrain=None, lr=0.0002, weight_decay=0.0001, num_epochs=500, iter_per_epoch=-1, region_fusion_start_epoch=20, seed=1024, beta_version=True, fusion_type='RFM', use_recon_loss=True, use_contrastive_loss=True, crop_size=112, name='hetero', train_transforms='Compose([RandCrop3D((112,112,112)), RandomRotion(10), RandomIntensityChange((0.1,0.1)), RandomFlip(0), NumpyType((np.float32, np.int64)),])', test_transforms='Compose([NumpyType((np.float32, np.int64)),])')

**Comment by CuCl-2** [comment]
Regarding the dataset split, there is a specific BraTS 2020 split file provided here. https://github.com/CuCl-2/DC-Seg/issues/4

**Comment by CuCl-2** [comment]
Data preprocessing can be done using this code.  https://github.com/dyh127/RFNet/blob/main/preprocess.py

**Comment by xiao-xinan** [comment]
感谢您的解答。

**Comment by seunghub** [comment]
感谢您的回复。我这边确认了一下，当 args.resume 为 True 时，需要将 crop_size=args.crop_size 作为参数传入 test_softmax。否则就会使用默认参数 80，从而得到 @xiao-xinan 所观察到的那种性能结果。

另外，由于单张 A6000 Ada GPU 显存放不下，我是使用 DataParallel 进行训练的，这可能也是性能无法成功复现的原因之一。同时，随机性（randomness）似乎也没有被完全控制住。请问您是否可以在如何控制 randomness 方面给一些建议呢？

**Comment by CuCl-2** [comment]
感谢您这个关于crop_size的发现，相信对大家复现非常有用。

DataParallel应该是有关系，如果batch_size为2的话，两个样本被分到了不同的gpu上，对于contrastive loss可能会造成影响。我用的是A800 80G，如果显存不够的话可以考虑crop_size降到80，用单卡跑一下看，可能性能相比于论文会有所下降，但应该是有限的。

代码中应该设置了random seed了

**Comment by Krystal024** [comment]
老师您好，我用batch size=1在一张5090训练需要58小时，这正常吗
